File: plugin-sass/src/main/java/com/jd/easy/audience/task/plugin/property/ConfigProperties.java
package com.jd.easy.audience.task.plugin.property;

import com.jd.easy.audience.common.constant.LabelSourceEnum;
import com.jd.easy.audience.common.constant.StringConstant;
import com.jd.easy.audience.common.exception.JobInterruptedException;
import com.jd.easy.audience.common.oss.OssClientTypeEnum;
import com.typesafe.config.Config;
import com.typesafe.config.ConfigFactory;
import org.apache.commons.lang.StringUtils;

import java.io.BufferedInputStream;
import java.io.IOException;
import java.io.InputStream;
import java.nio.charset.StandardCharsets;
import java.util.Base64;
import java.util.Properties;

/**
 * @author cdxiongmei
 * @title ConfigProperties
 * @description 配置文件中属性值
 * @date 2021/8/10 上午10:27
 * @throws
 */
public class ConfigProperties {
    /**
     * 属性对象
     */
    private static Properties properties;
    private static Config templateInfo;

    static {
        try {
            properties = getProperties();
            templateInfo = getTemplate();
        } catch (IOException e) {
            e.printStackTrace();
        }
    }

    /**
     * 对象存储协议模式
     */
    private static final String JSS_MODE = properties.getProperty("JSS_MODE");
    /**
     * 发布版本
     */
    public static final String VERSION = properties.getProperty("VERSION");
    /**
     * 4a流转统计周期分类-天流转
     */
    public static final String DAY_PERIOD = properties.getProperty("DAY_PERIOD");
    /**
     * 4a流转统计周期分类-周流转
     */
    public static final String WEEK_PERIOD = properties.getProperty("WEEK_PERIOD");
    /**
     * 4a流转统计周期分类-月流转
     */
    public static final String MONTH_PERIOD = properties.getProperty("MONTH_PERIOD");
    /**
     * 4a流转统计周期分类-七天流转
     */
    public static final String RECENT_7DAY_PERIOD = properties.getProperty("RECENT_7DAY_PERIOD");
    /**
     * 4a流转统计周期分类-30流转
     */
    public static final String RECENT_30DAY_PERIOD = properties.getProperty("RECENT_30DAY_PERIOD");
    /**
     * 标签数据集枚举值长度限制
     */
    public static final Integer LABEL_VALUE_LEN_MAX = Integer.valueOf(properties.getProperty("LABEL_VALUE_LEN_MAX"));
    /**
     * 标签数据集枚举值异常字符统一值
     */
    public static final String LABEL_VALUE_DEFAULT = "其他";
    /**
     * 人群生成状态
     */
    public static final String AUDIENCE_STATUS_KEY = "AudienceStatus";
    public static final String ONEID_DETAIL_TB = properties.getProperty("ONEID_DETAIL_TB");
//    private static final String ONE_ID_IDTYPE = "OneID";
    /**
     * 营销回执明细表
     */
    public static final String MARKETING_DETAIL = properties.getProperty("MARKETING_DETAIL");
    //数据库名称
    public static final String JED_DBNAME = properties.getProperty("JED_DBNAME");
    //数据库域名
    public static final String JED_URL = properties.getProperty("JED_URL");
    //数据库用户名
    public static final String JED_USER = properties.getProperty("JED_USER");
    //数据库密码
    public static final String JED_PASSWORD = new String(Base64.getDecoder().decode(properties.getProperty("JED_PASSWORD")), StandardCharsets.UTF_8);

    //ck配置
    //ck数据库名称
    public static final String CK_DBNAME = properties.getProperty("CK_DBNAME");
    //ck数据库域名
    public static final String CK_URL = properties.getProperty("CK_URL");
    //ck数据库用户名
    public static final String CK_USER = properties.getProperty("CK_USER");
    //ck数据库密码
    public static final String CK_PASS = properties.getProperty("CK_PASS");
    //ck数据表后缀
    public static final String CK_TABLESUFFIX = properties.getProperty("CK_TABLESUFFIX");
    //ck数据表前缀
    public static final String CK_TABLEPREFIX = properties.getProperty("CK_TABLEPREFIX");
    public static final String CK_SOCKET_TIMEOUT = properties.getProperty("CK_SOCKET_TIMEOUT");
    //应用端链接
    public static final String SERVICE_URI = properties.getProperty("SERVICE_URI");
    /**
     * 人群包标示类型为oneid
     */
    public static final Long ONEID_ONEIDTYPE = -99L;
    /**
     * oneid自动化应用中无法识别的用户标示类型
     */
    public static final Long ONEID_UNKNOWNTYPE = -1L;

    /**
     * @throws
     * @title getProperties
     * @description 获取属性值
     * @author cdxiongmei
     * @updateTime 2021/8/10 下午7:17
     * @return: java.util.Properties
     */
    private static Properties getProperties() throws IOException {
        Properties properties = new Properties();
        InputStream in = ConfigProperties.class.getResourceAsStream("/config.properties");
        properties.load(new BufferedInputStream(in));
        return properties;
    }

    /**
     * @throws
     * @title getClientType
     * @description 获取oss协议类型（从配置文件读取）
     * @author cdxiongmei
     * @updateTime 2021/8/10 下午7:18
     * @return: com.jd.easy.audience.common.oss.OssClientTypeEnum
     */
    public static OssClientTypeEnum getClientType() {
        if (StringUtils.isBlank(JSS_MODE)) {
            throw new JobInterruptedException("没有配置OSS 协议类型", "Please config your JSSMODE");
        }
        OssClientTypeEnum clientType = OssClientTypeEnum.valueOf(JSS_MODE);
        if (null == clientType) {
            throw new JobInterruptedException("配置OSS 协议类型不合法", "JSSMODE value is invalid");
        }
        return clientType;
    }

    public static String getLabelPrefix(LabelSourceEnum type) {
        return LabelSourceEnum.custom.equals(type) ? StringConstant.CUSTOM_LABELSETS_PREFIX : StringConstant.COMMON_LABELSETS_PREFIX;
    }

    private static Config getTemplate() {
        return ConfigFactory.parseResources("sqlTemplate.conf");
    }

    public static Config getsqlTemplate(String className, String methodName) {
        return ConfigProperties.templateInfo.getConfig(className).getConfig(methodName);
    }

    /**
     * 判断该用户标示是否需要oneid进行关联
     *
     * @param userIdType
     * @return
     */
    public static Boolean isNeedOneIdJoin(Long userIdType) {
        return !(ONEID_ONEIDTYPE.longValue() == userIdType.longValue() || ONEID_UNKNOWNTYPE.longValue() == userIdType.longValue());
    }
}

File: dataintegration-sass/src/main/scala/com/jd/easy/audience/task/dataintegration/property/ConfigProperties.scala
package com.jd.easy.audience.task.dataintegration.property

import com.jd.easy.audience.common.constant.NumberConstant

import java.io.{BufferedInputStream, BufferedReader, InputStream, InputStreamReader}
import java.nio.charset.StandardCharsets
import java.util.{Base64, Properties}
import com.jd.easy.audience.common.exception.JobInterruptedException
import com.jd.easy.audience.common.oss.OssClientTypeEnum
import com.typesafe.config.{Config, ConfigFactory}
import org.apache.commons.lang.StringUtils

/**
 * @title ConfigProperties
 * @description 全局配置属性
 * @author cdxiongmei
 * @updateTime 2021/8/9 下午9:12
 * @throws
 */
object ConfigProperties extends Serializable {

  private val properties = getProperties
  val templateInfo: Config = getTemplate
  //应用端链接
  val SERVICE_URI: String = properties.getProperty("SERVICE_URI")
  //对象存储协议类型
  val JSS_MODE: String = properties.getProperty("JSS_MODE")
  //数据库名称
  val JED_DBNAME: String = properties.getProperty("JED_DBNAME")
  //数据库域名
  val JED_URL: String = properties.getProperty("JED_URL")
  //数据库用户名
  val JED_USER: String = properties.getProperty("JED_USER")
  //数据库密码
  val JED_PASSWORD: String = new String(Base64.getDecoder.decode(properties.getProperty("JED_PASSWORD")), StandardCharsets.UTF_8)
  //ca数据拆分模版更新日志表
  val MARK_TABLE: String = properties.getProperty("MARK_TABLE")
  //cdp拆分结果日志
  val CDPLOG_TABLE: String = properties.getProperty("CDPLOG_TABLE")
  //dtshdfs path
  val DTS_HDFS_PATH: String = properties.getProperty("TARGET_BASE_PATH")
  val MODULE_DB_NAME: String = properties.getProperty("MODULE_DB_NAME")
  //  /**
  //   * hdfs保存营销回执明细
  //   */
  private val WRITER_BATCH_PATH_MODULE: String = properties.getProperty("WRITER_BATCH_PATH")
  //  /**
  //   * 营销回执明细
  //   */
  //  val MARKETING_DETAIL_TB: String = properties.getProperty("MARKETING_DETAIL")
  val TABLE_INFO_TB: String = "ea_create_table_info"
  val CDP_TABLE_PREFIX: String = "adm_"
  val BDP_USER: String = "USER"
  val HADOOP_USER: String = "HADOOP_USER_NAME"
  val BUFFALO_ENV_NTIME: String = "BUFFALO_ENV_NTIME"
  val BUFFALO_ENV_BCYCLE: String = "BUFFALO_ENV_BCYCLE" //2021-09-16-21-15
  val BUFFALO_ENV_CYCLE_TYPE: String = "BUFFALO_ENV_CYCLE_TYPE" //minute,hour,day,week
  val BUFFALO_ENV_YDAY: String = "BUFFALO_ENV_YDAY" //系统前一天时间（日期）yyyyMMdd
  val PUBLIC_DB_NAME: String = "ae_public"
  val PUBLIC_DB_NAME_DB: String = "ae_public.db"
  val DATABASE_URI: String = "hdfs://ns1/user/cdp_biz-org.bdp.cs"
  val BUILD_SPLIT_BEANS_CLASS: String = "buildBeansClass"
  val FILTER_PATITION_FIELD: String = "isPartition = 'true'"
  val FILTER_ID_NOT_NULL: String = "id is not null"
  val FIELD_ID: String = "id"
  val FIELD_ACCOUNT_ID: String = "account_id"
  val FIELD_MAIN_ACCOUNT_NAME: String = "main_account_name"
  val FIELD_SUB_ACCOUNT_NAME: String = "sub_account_name"
  val FIELD_DATABASE_NAME: String = "database_name"
  val FIELD_TABLE_DATA_UPDATE_TIME: String = "table_data_update_time"
  val FIELD_TABLE_NAME: String = "table_name"
  val FIELD_TABLE_COMMENT_NEW: String = "table_comment_new"
  val FIELD_COL_NAME: String = "col_name"
  val FIELD_LOCATION: String = "Location"
  val FIELD_DATA_TYPE: String = "data_type"
  val DTS_FIELD_TABLENAME: String = "@tableName@"
  val DATEFORMATE_YYYMMDD_HH: String = "yyyy-MM-dd/HH"
  val DATEFORMATE_YYYMMDD_HH_MM: String = "yyyy-MM-dd/HH/mm"
  val DATEFORMATE_YYYMMDD: String = "yyyy-MM-dd"
  val DATEFORMATE_TIMESTAMP: String = "yyyy-MM-dd HH:mm:ss"
  val UPDATE_LIMITNUM: Int = NumberConstant.INT_40



  private def getProperties: Properties = {
    val properties: Properties = new Properties
    val in: InputStream = getClass.getResourceAsStream(s"/config.properties")
    properties.load(new BufferedInputStream(in))
    properties
  }

  private def getTemplate: Config = {
    ConfigFactory.parseResources(s"template.conf")
  }

  def getClientType: OssClientTypeEnum = {
    if (StringUtils.isBlank(JSS_MODE)) throw new JobInterruptedException("没有配置OSS 协议类型", "The value of JSS_Mode is not exist!")
    val clientType = OssClientTypeEnum.valueOf(JSS_MODE)
    if (null == clientType) throw new JobInterruptedException("配置OSS 协议类型不合法", "The value of JSS_Mode is illegal！")
    clientType
  }

  def getHdfsPath(bdpUser: String): String = {
    String.format(WRITER_BATCH_PATH_MODULE, bdpUser)
  }

  def getCreateTableModule(dbName: String, tableName: String): String = {
    val in: InputStream = getClass.getResourceAsStream(s"/createTableModule/" + tableName)
    val br = new BufferedReader(new InputStreamReader(in))
    var result = new StringBuilder
    result = result.append(Stream.continually(br.readLine()).takeWhile(_ != null).mkString("\n"))
    String.format(result.toString(), dbName)
  }


}
